src/analyser.cpp

```diff
@@ -0,0 +1,1509 @@
+                }
+            }
+        }
+
+        if (((astParent->mPimpl->mType != AnalyserEquationAst::Type::ASSIGNMENT)
+             || (astParent->mPimpl->mLeftChild != ast))
+            && (astParent->mPimpl->mType != AnalyserEquationAst::Type::BVAR)) {
+            // We are dealing with a variable which is neither a computed
+            // variable nor our variable of integration, so retrieve its scaling
+            // factor and apply it, if needed, distinguishing between a rate
+            // variable and an algebraic variable.
+
+            auto scalingFactor = Analyser::AnalyserImpl::scalingFactor(ast->mPimpl->mVariable);
+
+            if (!areEqual(scalingFactor, 1.0)) {
+                if (astParent->mPimpl->mType == AnalyserEquationAst::Type::DIFF) {
+                    scaleAst(astParent, astParent->mPimpl->mParent.lock(), scalingFactor);
+                } else {
+                    scaleAst(ast, astParent, scalingFactor);
+                }
+            }
+        }
+    }
+}
+
+void Analyser::AnalyserImpl::processModel(const ModelPtr &model)
+{
+    // Reset a few things in case this analyser was to be used to process more
+    // than one model.
+
+    mAnalyser->removeAllIssues();
+
+    mModel = std::shared_ptr<AnalyserModel> {new AnalyserModel {}};
+
+    mInternalVariables.clear();
+    mInternalEquations.clear();
+
+    // Recursively process the model's components, so that we end up with an AST
+    // for each of the model's equations.
+
+    for (size_t i = 0; i < model->componentCount(); ++i) {
+        processComponent(model->component(i));
+    }
+
+    // Some more processing is needed, but it can only be done if we didn't come
+    // across any errors during the processing of our components.
+
+    if (mAnalyser->errorCount() == 0) {
+        // Process our different equations' AST to determine the type of our
+        // variables.
+
+        for (const auto &internalEquation : mInternalEquations) {
+            processEquationAst(internalEquation->mAst);
+        }
+    }
+
+    // Some post-processing is now needed, but it can only be done if we didn't
+    // come across any errors during the processing of our equations' AST.
+
+    if (mAnalyser->errorCount() == 0) {
+        // Sort our variables, determine the index of our constant variables and
+        // then loop over our equations, checking which variables, if any, can
+        // be determined using a given equation.
+
+        std::sort(mInternalVariables.begin(), mInternalVariables.end(),
+                  compareVariablesByName);
+
+        auto variableIndex = MAX_SIZE_T;
+
+        for (const auto &internalVariable : mInternalVariables) {
+            if (internalVariable->mType == AnalyserInternalVariable::Type::CONSTANT) {
+                internalVariable->mIndex = ++variableIndex;
+            }
+        }
+
+        auto equationOrder = MAX_SIZE_T;
+        auto stateIndex = MAX_SIZE_T;
+
+        for (;;) {
```

Comment:
Is it just personal preference whether to use a while loop for `relevantCheck` or an infinite for loop like this? Is there any style guide?

Reply:
Just how it came to me when I wrote that piece of code. However, I can see that we don't use `for (;;)` anywhere else, so I have replaced it with:
```
        bool relevantCheck;

        do {
            relevantCheck = false;

            for (const auto &internalEquation : mInternalEquations) {
                relevantCheck = internalEquation->check(equationOrder, stateIndex, variableIndex)
                                || relevantCheck;
            }
        } while (relevantCheck);
```

Done in commit dc6c820a.